File: client/native/share.js
// native/share.js — Web Share API glue for the N2 share capability.
//
// `navigator.share` requires a user gesture and a secure context. When the API
// is missing, share() returns `{ outcome: "unsupported" }` rather than throwing,
// so callers fall back gracefully (e.g. to the clipboard). A user dismissal
// resolves `{ outcome: "cancelled" }`.

/**
 * Report whether the Web Share API is available.
 * @param {Object} _args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{supported:boolean}>}
 */
export async function shareIsSupported(_args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  return { supported: Boolean(nav && typeof nav.share === "function") };
}

/**
 * Open the OS share sheet, degrading gracefully when unsupported.
 * @param {{title:string,text:string,url:string,files:Array<Object>}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{outcome:"shared"|"cancelled"|"unsupported"}>}
 */
export async function shareShare(args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  if (!nav || typeof nav.share !== "function") {
    return { outcome: "unsupported" };
  }
  /** @type {ShareData} */
  const data = {};
  if (args.title) data.title = args.title;
  if (args.text) data.text = args.text;
  if (args.url) data.url = args.url;
  // File sharing has uneven support; only forward when the browser accepts it.
  if (Array.isArray(args.files) && args.files.length > 0 && nav.canShare) {
    const files = args.files;
    if (nav.canShare({ files })) data.files = /** @type {any} */ (files);
  }
  try {
    await nav.share(data);
    return { outcome: "shared" };
  } catch (err) {
    // AbortError means the user dismissed the sheet — a normal outcome.
    if (err && err.name === "AbortError") return { outcome: "cancelled" };
    return { outcome: "cancelled" };
  }
}

File: tempestweb/native/share.py
"""Native share capability over the browser's Web Share API (N2).

The pythonic mirror of the React SDK's ``share`` / ``isShareSupported``. Opens the
OS share sheet via ``navigator.share`` when available, and degrades gracefully
otherwise. Pairs with the ``share_target`` manifest field (P5).

``client/native/share.js`` calls ``navigator.share`` (a user gesture and a secure
context are required). When the API is missing it returns an ``unsupported``
outcome rather than throwing, so callers can fall back (e.g. to the clipboard).
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict

from tempestweb.native.dispatch import send_native_call

__all__ = ["ShareOutcome", "ShareResult", "is_share_supported", "share"]


class ShareOutcome(StrEnum):
    """The outcome of a :func:`share` call.

    Attributes:
        SHARED: The OS share sheet completed (content was shared).
        CANCELLED: The user dismissed the share sheet.
        UNSUPPORTED: The Web Share API is unavailable in this browser.
    """

    SHARED = "shared"
    CANCELLED = "cancelled"
    UNSUPPORTED = "unsupported"


class ShareResult(BaseModel):
    """The typed result of a :func:`share` call.

    Attributes:
        outcome: The :class:`ShareOutcome`.
    """

    model_config = ConfigDict(frozen=True)

    outcome: ShareOutcome


async def is_share_supported() -> bool:
    """Report whether the Web Share API is available in the current browser.

    Returns:
        ``True`` if ``navigator.share`` exists (and the context permits sharing),
        ``False`` otherwise.

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("share.is_supported", {})
    return bool(value.get("supported", False))


async def share(
    *,
    title: str = "",
    text: str = "",
    url: str = "",
    files: list[dict[str, str]] | None = None,
) -> ShareResult:
    """Open the OS share sheet, falling back gracefully when unsupported.

    Args:
        title: The share title.
        text: The share body text.
        url: A URL to share.
        files: Optional file descriptors (``{"name", "type", "data"}``); file
            sharing has uneven browser support and may downgrade to text/url.

    Returns:
        A :class:`ShareResult`. ``ShareOutcome.UNSUPPORTED`` and
        ``ShareOutcome.CANCELLED`` are normal, non-error outcomes — never raised.

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call(
        "share.share",
        {"title": title, "text": text, "url": url, "files": files or []},
    )
    outcome = ShareOutcome(str(value.get("outcome", ShareOutcome.UNSUPPORTED.value)))
    return ShareResult(outcome=outcome)
